lib/improver/weights.py

```diff
@@ -30,123 +30,259 @@
-    if cval <= 0.0 or cval > 1.0:
-        msg = ('cval must be greater than 0.0 and less '
-               'than or equal to 1.0 '
-               'cval = {0:s}'.format(str(cval)))
-        raise ValueError(msg)
-
-    weights_list = []
-    for tval_minus1 in range(0, num_of_weights):
-        weights_list.append(cval**(tval_minus1))
-
-    weights = normalise_weights(np.array(weights_list))
-
-    return weights
-
-
-def linear_weights(num_of_weights, y0val=1.0, slope=0.0,
-                   ynval=None):
-    """Create linear weights
-
-        Args:
-            num_of_weights : Positive Integer
-                             Number of weights to create.
-            y0val : Positive float
-                    relative value of starting point. Default = 1.0
-
-            AND EITHER:
-            slope : float
-                    slope of the line. Default = 0.0 (equal weights)
-            OR
-            ynval : Positive float or None
-                    Relative weights of last point.
-                    Default value is None
-
-        Returns:
-            weights : array of weights
-                      sum of all weights = 1.0
-
-    """
-    if not isinstance(num_of_weights, int) or num_of_weights <= 0:
-        msg = ('Number of weights must be integer > 0 '
-               'num = {0:s}'.format(str(num_of_weights)))
-        raise ValueError(msg)
-    # Special case num_of_weighs == 1 i.e. Scalar coordinate.
-    if num_of_weights == 1:
-        weights = np.array([1.0])
-        return weights
-    if not isinstance(y0val, float) or y0val <= 0.0:
-        msg = ('y0val must be a float > 0.0, '
-               'y0val = {0:s}'.format(str(y0val)))
-        raise ValueError(msg)
-    if ynval is not None:
-        if slope != 0.0:
-            msg = ('Relative end point weight or slope must be set'
-                   ' but not both.')
+class WeightsUtilities(object):
+    """ Utilities for Weight processing. """
+
+    def __init__(self):
+        """Initialise class."""
+        pass
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<WeightsUtilities>')
+        return result
+
+    @staticmethod
+    def normalise_weights(weights):
+        """Ensures all weights add up to one.
+
+            Args:
+                weights : array of weights.
+
+            Returns:
+                normalised_weights : array of weights
+                                     where sum = 1.0
+        """
+        if weights.min() < 0.0:
+            msg = 'Weights must be positive, at least one value < 0.0'
+            raise ValueError(msg)
+
+        sumval = weights.sum()
+        if sumval == 0:
+            msg = 'Sum of weights must be > 0.0'
             raise ValueError(msg)
+
+        normalised_weights = weights / sumval
+        return normalised_weights
+
+    @staticmethod
+    def redistribute_weights(weights, missing_weights, method='evenly'):
```

Comment:
Added to the CLI. There are too many arguments in the CLI. It isn't simple and it isn't clear.